src/httpclient.c

```diff
@@ -230,8 +242,84 @@ int httpclient_log(
     corto_dealloc(bytes);
     */
     corto_buffer_appendstr(&g_logBuffer, data);
-
     // corto_trace("%s", corto_buffer_str(&buffer));
+    return 0;
+}
+
+corto_string httpclient_encodeFields(
+    corto_string fields)
+{
+    corto_string encoded = curl_easy_escape(NULL, fields, 0);
+    return encoded;
+}
+
+/* Maximum time in milliseconds that you allow the libcurl transfer operation
+ * to take. Normally, name lookups can take a considerable time and limiting
+ * operations to less than a few minutes risk aborting perfectly normal
+ * operations. This option may cause libcurl to use the SIGALRM signal to
+ * timeout system calls.
+ */
+int16_t httpclient_setTimeout(
+    int32_t timeout)
+{
+    if (corto_threadTlsKey(&TLS_TimeoutKey, NULL)) {
+        corto_seterr("Failed to initialize timeout key. Error: %s",
+            corto_lasterr());
+        goto error;
+    }
+
+    if (corto_threadTlsSet(TLS_TimeoutKey, (void *)(intptr_t)timeout)) {
+        corto_seterr("Failed to set TLS timeout data. %s", corto_lasterr());
+        goto error;
+    }
+
+    return 0;
+error:
+    return -1;
+}
+
+int32_t httpclient_getTimeout(void)
+{
+    int32_t timeout = DEFAULT_TIMEOUT;
+
+    void *data = corto_threadTlsGet(TLS_TimeoutKey);
+    if (data) {
+        timeout = (int32_t)(intptr_t)data;
+    }
+
+    return timeout;
+}
+
+/* Maximum time, in milliseconds, that the connection phase is allowed to
+ * execute before failing to connect to host */
+int16_t httpclient_setConnectTimeout(
+    int32_t timeout)
+{
+    if (corto_threadTlsKey(&TLS_ConnectTimeoutKey, NULL)) {
```

Comment:
This doesn't work- you need to generate the key once with `corto_threadTlsKey` (usually done in the package main). Otherwise you will keep generating new keys every time you set a timeout. This won't work in multithreaded contexts, since a new thread can overwrite a key, which will cause other threads to no longer being able to obtain their value.